feat: Editing flow

NidhiKJha opened the pull request:
This PR is to add editing flow while switching networks via dapp
This PR also includes the connections flow which includes editing


## **Related issues**

Fixes: 
1. https://github.com/MetaMask/MetaMask-planning/issues/2683
2. https://github.com/MetaMask/MetaMask-planning/issues/2684
3. https://github.com/MetaMask/MetaMask-planning/issues/2697
4. https://github.com/MetaMask/MetaMask-planning/issues/2698
5. https://github.com/MetaMask/MetaMask-planning/issues/2663

## **Manual testing steps**

1. run the extension with CHAIN_PERMISSIONS flag
2. Connect to a dapp 
3. Switch network via metamask
6. toast should show up
7. Then click on edit permissions, it should route to new permissions screen
8. We have edit networks and edit accounts modal there

## **Screenshots/Recordings**


### **Before**

NA
### **After**


https://github.com/user-attachments/assets/a9dbaecf-c96b-485c-b639-f50cbe30317c

https://github.com/user-attachments/assets/8638bb1f-3739-4a2b-84ae-dca24f8abe6d



## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

NidhiKJha committed:
added networks screen modal

NidhiKJha committed:
added test and checkbo

NidhiKJha committed:
lint fix

NidhiKJha committed:
lint fix

NidhiKJha committed:
updated snapshot

NidhiKJha committed:
updated jest

NidhiKJha committed:
updated tests

NidhiKJha committed:
lint fix

NidhiKJha committed:
added edit accounts modal

NidhiKJha committed:
added toast for permission switch

NidhiKJha committed:
added review perm page

NidhiKJha committed:
added review permission page

NidhiKJha committed:
editing flow

NidhiKJha committed:
updated selector

adonesky1 committed:
fix

adonesky1 commented on ui/selectors/permissions.js:
```diff
@@ -256,13 +269,37 @@ function getAccountsPermissionFromSubject(subject = {}) {
+}
+
+function getChainsPermissionFromSubject(subject = {}) {
+  return subject.permissions?.['endowment:permitted-chains'] || {};
```
lets use 
```suggestion
  return subject.permissions?.[PermissionNames.permittedChains] || {};
```

NidhiKJha committed:
permitted chains and accounts updated for review screen

NidhiKJha committed:
removed unused console

NidhiKJha committed:
added edit accounts functionality

NidhiKJha committed:
updated toast

NidhiKJha committed:
added no connection page

NidhiKJha committed:
added edit networks functionality

adonesky1 committed:
fix

NidhiKJha committed:
removed console statements

NidhiKJha committed:
updated final screen for editing flow

NidhiKJha committed:
updated controllers

NidhiKJha committed:
updated edit networks

NidhiKJha committed:
lint fix

NidhiKJha committed:
lint fix

NidhiKJha committed:
prep build fix

NidhiKJha committed:
lint fix

adonesky1 commented:
Ok so seeing a few issues when testing it out, here's one and will post more videos

See here how the permissions are not reflected accurately by the checkboxes? 
https://github.com/user-attachments/assets/cc864093-543e-41ef-8609-d4cf6cdb9e00

jonybur commented on ui/components/app/permission-page-container/permission-page-container.component.js:
```diff
@@ -146,10 +146,11 @@ export default class PermissionPageContainer extends Component {
-        approvedChainIds: _request.permissions?.permittedChains?.caveats.find(
-          (caveat) => caveat.type === 'restrictNetworkSwitching',
-        )?.value,
+      ...(_request.permissions[PermissionNames.permittedChains] && {
```
I think for the sake of readability we could split this one-liner logic into several lines of code with named variables

jiexi replied:
good point. done here https://github.com/MetaMask/metamask-extension/pull/26635/commits/0dc84aafa710174fa4ad9569a18bf965ce3e45e7

jonybur commented on ui/components/multichain/edit-accounts-modal/edit-accounts-modal.tsx:
```diff
@@ -17,65 +22,180 @@ import {
+
+  const activeTabOrigin = useSelector(getOriginOfCurrentTab);
+  const subjects = useSelector(getPermissionSubjects);
+  const connectedAccounts = useSelector((state: any) =>
```
If possible let's avoid the use of `any`

NidhiKJha replied:
Replaced it with   const connectedAccounts = useSelector((state) =>
    getOrderedConnectedAccountsForConnectedDapp(state, activeTabOrigin).filter(
      (account: InternalAccount) => isEvmAccountType(account.type),
    ),
  );

jonybur commented on ui/components/multichain/edit-accounts-modal/edit-accounts-modal.tsx:
```diff
@@ -17,65 +22,180 @@ import {
+        dispatch(
+          removePermissionsFor(
+            permissionsRecord as Record<string, NonEmptyArray<string>>,
+          ),
```
I see a lot of type casting in this method, which can be a code smell. Can we infer the types?

NidhiKJha replied:
Updated it

jiexi replied:
unfortunately the `.length > 0` check doesn't help typescript infer the NonEmptyArray. I've removed most of the type casts but had to leave that one

https://github.com/MetaMask/metamask-extension/pull/26635/commits/8984bab4d07d3cf5ceb8ba43d046b608a9b81bc9

jonybur commented on ui/components/multichain/edit-networks-modal/edit-networks-modal.js:
```diff
@@ -12,13 +17,75 @@ import {
+  } else if (selectedChains.length > 0 && !allAreSelected()) {
+    checked = false;
+    isIndeterminate = true;
+  }
```
Can be simplified like this:

```
const checked = allAreSelected();
const isIndeterminate = selectedChains.length > 0 && !checked;
```

NidhiKJha replied:
Updated the logic and we need it to allAreSelected, with checked it doesn't seem to work

jiexi replied:
https://github.com/MetaMask/metamask-extension/pull/26635/commits/d065a0cb67631ccf420c3d27c1896ff9552cbc3e

jonybur commented on ui/components/multichain/edit-networks-modal/edit-networks-modal.js:
```diff
@@ -65,12 +134,47 @@ export const EditNetworksModal = ({ onClose }) => {
+            <ButtonPrimary
+              data-testid="disconnect-chains-button"
+              onClick={() => {
+                // disconnectAllAccounts();
```
Unused method

NidhiKJha replied:
Updated and passed via prop

jonybur commented on ui/components/multichain/pages/review-permissions-page/review-permissions-page.tsx:
```diff
@@ -0,0 +1,204 @@
+  );
+  const subjects = useSelector(getPermissionSubjects);
+  const grantedNetworks = networksList.filter(
+    (net: { chainId: any }) => connectedNetworks.indexOf(net.chainId) !== -1,
```
Use of `any`

NidhiKJha replied:
Removed

jonybur commented on ui/components/multichain/pages/review-permissions-page/review-permissions-page.tsx:
```diff
@@ -0,0 +1,204 @@
+    (net: { chainId: any }) => connectedNetworks.indexOf(net.chainId) !== -1,
+  );
+  const hostName = getURLHost(securedOrigin);
+  const disconnectAllAccounts = () => {
```
This method is also present in `edit-accounts-modal.tsx`, line 93

NidhiKJha replied:
It's removed from edit-accounts-modal.tsx now. Passed via this component now

jonybur commented on ui/components/multichain/pages/review-permissions-page/review-permissions-page.tsx:
```diff
@@ -0,0 +1,204 @@
+                iconName={IconName.ArrowLeft}
+                className="connections-header__start-accessory"
+                size={ButtonIconSize.Sm}
+                onClick={() => (history as any).goBack()}
```
Use of `any`

NidhiKJha replied:
I couldn't determine history here. So, it's fine to use any here

jonybur commented on ui/components/multichain/pages/review-permissions-page/site-cell/site-cell.tsx:
```diff
@@ -0,0 +1,224 @@
+        alignItems={AlignItems.baseline}
+        width={BlockSize.Full}
+        backgroundColor={BackgroundColor.backgroundDefault}
+        // onClick={onClick}
```
Unused onClick

NidhiKJha replied:
It's updated

jonybur commented on ui/components/multichain/pages/review-permissions-page/site-cell/site-cell.tsx:
```diff
@@ -0,0 +1,224 @@
+        alignItems={AlignItems.baseline}
+        width={BlockSize.Full}
+        backgroundColor={BackgroundColor.backgroundDefault}
+        // onClick={onClick}
```
Unused onClick

NidhiKJha replied:
Removed

bergeron commented on ui/components/multichain/network-list-menu/network-list-menu.js:
```diff
@@ -303,15 +310,20 @@ export const NetworkListMenu = ({ onClose }) => {
-          // If presently on and connected to a dapp, communicate a change to
-          // the dapp via silent switchEthereumChain that the network has
-          // changed due to user action
+          if (process.env.CHAIN_PERMISSIONS) {
```
Can you share the context for what this does?  This file is rewritten in the default RPC work and I want to make sure this bit gets integrated correctly

georgewrmarshall commented on ui/components/multichain/edit-accounts-modal/edit-accounts-modal.tsx:
```diff
@@ -17,65 +22,180 @@ import {
       <ModalOverlay />
       <ModalContent>
         <ModalHeader onClose={onClose}>{t('editAccounts')}</ModalHeader>
         <Box padding={4}>
```
Hey @NidhiKJha, wrapping the body components in `ModalBody` should fix the issues here

```suggestion
        <ModalBody paddingLeft={0} paddingRight={0}>
          <Box padding={4}>
```

darkwing commented on ui/components/multichain/edit-accounts-modal/edit-accounts-modal.tsx:
```diff
@@ -17,65 +22,180 @@ import {
+        <ModalFooter>
+          {selectedAccounts.length === 0 ? (
+            <ButtonPrimary
+              data-testid="connect-more-accounts-button"
```
Would it be helpful to have a separate value here?  `disconnect-more-accounts-button`?

NidhiKJha replied:
Updated it

darkwing commented on ui/components/multichain/edit-accounts-modal/edit-accounts-modal.tsx:
```diff
@@ -17,65 +22,180 @@ import {
+    connectedAccountsAddresses: string[],
+    activeTabOrigin: string,
+  ) => {
+    const removedElements = connectedAccountsAddresses.filter(
```
Should any of this data handling be done in a helper function or action/selector?

NidhiKJha replied:
Removed this logic

NidhiKJha commented:
@jonybur thanks for the review. I have tried to address all the comments in this PR https://github.com/MetaMask/metamask-extension/pull/26922 as it's where we finalised most of the types. It's in draft state but will address all your comments in that one

NidhiKJha committed:
feat: Connections flow (#26922)

This PR is to add connections Flow

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask
Extension Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

---------

Co-authored-by: Alex <adonesky@gmail.com>

NidhiKJha committed:
Merge branch 'develop' into editing-flow

NidhiKJha committed:
lint fix

NidhiKJha committed:
lint fix

NidhiKJha committed:
error fix

NidhiKJha committed:
updated permissions

NidhiKJha committed:
jest update

NidhiKJha committed:
lint fix

darkwing commented on ui/pages/permissions-connect/connect-page/connect-page.tsx:
```diff
@@ -0,0 +1,148 @@
+    >
+      <Header>
+        <Text variant={TextVariant.headingLg}>{t('connectWithMetaMask')}</Text>
+        <Text>This site wants to: </Text>
```
Needs localization.

NidhiKJha replied:
Replaced it

darkwing commented on ui/pages/permissions-connect/connect-page/connect-page.tsx:
```diff
@@ -0,0 +1,148 @@
+          networks={filteredNetworks}
+          accounts={filterAccountsByAddress}
+          onAccountsClick={() => selectAccounts(Array.from(selectedAccounts))}
+          onNetworksClick={() => console.log('testing')}
```
`console.log` needs removal.

NidhiKJha replied:
Removed

darkwing commented on ui/pages/permissions-connect/connect-page/connect-page.tsx:
```diff
@@ -0,0 +1,148 @@
+    : networksList;
+
+  // Select approved accounts and networks
+  const approvedAccounts =
```
Could this be a selector?

NidhiKJha replied:
We can move it in a selector but since it's based on just one condition, and we are using it one file. We can keep it here

darkwing commented on ui/pages/permissions-connect/connect-page/connect-page.tsx:
```diff
@@ -0,0 +1,148 @@
+    setSelectedAccounts(newSelectedAccounts);
+  };
+  // Filter networks based on chainId
+  const filteredNetworks = Array.isArray(selectednetworksList)
```
Could this be a selector?

NidhiKJha replied:
This is updated

darkwing commented on ui/components/multichain/pages/review-permissions-page/review-permissions-page.tsx:
```diff
@@ -0,0 +1,298 @@
+              </ButtonPrimary>
+            )}
+          </>
+        </Footer>{' '}
```
Is the extra space necessary?

NidhiKJha replied:
Removed

darkwing commented on ui/components/multichain/pages/review-permissions-page/review-permissions-page.tsx:
```diff
@@ -0,0 +1,298 @@
+              className="connections-header__start-accessory"
+              size={ButtonIconSize.Sm}
+              // eslint-disable-next-line @typescript-eslint/no-explicit-any
+              onClick={() => (history as any).goBack()}
```
Do we know what type this could be?  `Window.history`?  Does this need to be variable or do we know where this should go every time?

NidhiKJha replied:
This goes back to the page where it was previously

darkwing commented on ui/components/multichain/pages/review-permissions-page/review-permissions-page.tsx:
```diff
@@ -0,0 +1,298 @@
+      );
+      history.push(`${CONNECT_ROUTE}/${requestId}`);
+    } else {
+      console.error('Tab origin is null.');
```
Should we throw an error here?  I don't know that `console.error` is helpful.

NidhiKJha replied:
This shouldn't throw an error. Because it's just a simple route. Remove the else condition

NidhiKJha committed:
updated lint errors

NidhiKJha committed:
lint fix

jiexi commented on app/scripts/metamask-controller.js:
```diff
@@ -5567,11 +5574,26 @@ export default class MetamaskController extends EventEmitter {
+              },
+            },
+          }),
+        requestPermissionsForOrigin: (requestedPermissions) => {
           this.permissionController.requestPermissions.bind(
             this.permissionController,
             { origin },
```
this shouldn't be a `.bind` anymore and also needs to return the value of the call

```suggestion
        requestPermissionsForOrigin: (requestedPermissions) => {
          return this.permissionController.requestPermissions(
            { origin },
```

adonesky1 commented on app/scripts/lib/rpc-method-middleware/handlers/ethereum-chain-utils.js:
```diff
@@ -207,7 +209,11 @@ export async function switchChain(
         permissionedChainIds === undefined ||
         !permissionedChainIds.includes(chainId)
       ) {
-        await requestPermittedChainsPermission([chainId]);
+        if (isAddFlow) {
+          await grantPermittedChainsPermissionIncremental([chainId]);
+        } else {
+          await requestPermittedChainsPermission([chainId]);
+        }
```
that's correct

NidhiKJha committed:
lint fix

NidhiKJha committed:
updated controller

NidhiKJha committed:
Merge branch 'develop' into editing-flow

NidhiKJha committed:
updated syntax

NidhiKJha committed:
updated stories

NidhiKJha committed:
lint fix

jiexi committed:
Fix requestPermissionsForOrigin

jiexi committed:
Merge remote-tracking branch 'origin/editing-flow' into editing-flow

jiexi committed:
only request permittedChains when feature flag and eth_accounts is requested too

jiexi committed:
remove approvedChainIds from legacy permission page submit

jiexi committed:
Merge branch 'develop' into editing-flow

jiexi committed:
Merge branch 'develop' into editing-flow

jiexi committed:
Jl/editing flow fix incremental (#27181)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/27181?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask
Extension Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

NidhiKJha committed:
updated edit accounts modal story

NidhiKJha committed:
lint fix

jiexi committed:
Fix wallet_addEthereumChain handler test

jiexi committed:
Merge branch 'develop' into editing-flow

NidhiKJha committed:
updated changes

NidhiKJha committed:
lint fix

jiexi committed:
Jl/editing flow network client switch (#27225)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

Switch the selected network for the domain and the globally selected
network when the permittedChains permission changes and the currently
selected network for the domain is no longer in the list of permitted
chains

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/27225?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask
Extension Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

NidhiKJha committed:
Merge branch 'develop' into editing-flow

NidhiKJha committed:
fixed ts errors

jiexi committed:
Fix permissions/selectors

NidhiKJha committed:
fixed ts errors

NidhiKJha committed:
Merge branch 'develop' into editing-flow

NidhiKJha committed:
Merge branch 'develop' into editing-flow

NidhiKJha committed:
lint fix

jiexi committed:
Fix sonar cloud default assignment warning

jiexi committed:
remove infura key

jiexi committed:
Merge branch 'develop' into editing-flow

jiexi committed:
Merge branch 'develop' into editing-flow

jiexi committed:
update data-testid tags

jiexi committed:
Merge branch 'develop' into editing-flow

jiexi committed:
Deal with code dupe

jiexi committed:
Merge branch 'develop' into editing-flow

NidhiKJha committed:
Merge branch 'develop' into editing-flow

NidhiKJha committed:
removed yarn change

NidhiKJha committed:
updated url Params

NidhiKJha committed:
updated url Params

adonesky1 commented:
A design/copy question:
<img width="360" alt="Screenshot 2024-09-18 at 2 04 09 PM" src="https://github.com/user-attachments/assets/33151582-1724-465c-bfd2-0f765a725e11">


Should the disconnect button now read "Disconnect all accounts and networks"? 

Same question with the manual "connect accounts" button:
<img width="358" alt="Screenshot 2024-09-18 at 2 05 22 PM" src="https://github.com/user-attachments/assets/62e8981b-98c0-4e33-b9e0-a0a3fc58a2c5">

Should it be "connect accounts and networks"?

adonesky1 commented on app/scripts/controllers/permissions/selectors.js:
```diff
@@ -40,47 +41,74 @@ export const getPermittedAccountsByOrigin = createSelector(
-export const getChangedAccounts = (newAccountsMap, previousAccountsMap) => {
-  if (previousAccountsMap === undefined) {
-    return newAccountsMap;
+export const getChangedOrigins = (newOriginsMap, previousOriginsMap) => {
```
I'm a bit confused by the use of the term `origins` here? Should it be `getChangedPermissions`? I'm not seeing how it relates to origins at all? Actually it really even more generic than `permissions` its sortof a pure agnostic comparison helper function no?

adonesky1 replied:
`getChangedValues`?

jiexi replied:
renamed this helper to `diffMap` and updated the internal variable names https://github.com/MetaMask/metamask-extension/pull/26635/commits/024597d47d18f6857ec23998e6b5124a41c130bb

adonesky1 commented on app/scripts/controllers/permissions/selectors.test.js:
```diff
@@ -113,4 +119,89 @@ describe('PermissionController selectors', () => {
   });
+
+  describe('getPermittedChainsByOrigin', () => {
+    it('memoizes and gets permitted chains by origin', () => {
```
How is it memoizing them? I'm not seeing that in the selector?

jiexi replied:
i just copied this from the other scenario, but 

https://github.com/reduxjs/reselect#:~:text=Reselect%20exports%20a%20createSelector%20API%2C%20which%20generates%20memoized%20selector%20functions

adonesky1 commented on ui/components/multichain/network-list-menu/network-list-menu.js:
```diff
@@ -171,7 +176,9 @@ export const NetworkListMenu = ({ onClose }) => {
     return sortedNonTestNetworks;
   };
 
+  // check if not granted chain is clicked, then show the toast and grant permission
```
[nit] a bit unclear what this comment is pointing to

NidhiKJha replied:
It's removed now here https://github.com/MetaMask/metamask-extension/pull/26635/commits/3868d34057e993b1cbcdf570bd5765b62e4bf5e7. I added it initially for myself as a TODO

adonesky1 commented on ui/components/multichain/network-list-menu/network-list-menu.js:
```diff
@@ -303,15 +310,20 @@ export const NetworkListMenu = ({ onClose }) => {
+          if (process.env.CHAIN_PERMISSIONS) {
+            grantPermittedChain(selectedTabOrigin, network.chainId);
+            if (!flattenedPermittedChains.includes(network.chainId)) {
+              dispatch(showPermittedNetworkToast());
```
I don't believe I'm seeing this toast on latest?

adonesky1 replied:
https://github.com/user-attachments/assets/ca492a27-65a0-4809-b279-84679248f816

NidhiKJha replied:
This toast is supposed to show up on the home page when the user grants permission to an enabled network. It seems to be working for me

https://github.com/user-attachments/assets/26243db8-94dc-490e-9870-6a5f6e907c24

adonesky1 replied:
hmmm yeah I expected to see it there but did not

adonesky1 commented on ui/selectors/permissions.js:
```diff
@@ -256,13 +281,37 @@ function getAccountsPermissionFromSubject(subject = {}) {
+}
+
+function getChainsPermissionFromSubject(subject = {}) {
+  return subject.permissions?.['endowment:permitted-chains'] || {};
```
```suggestion
  return subject.permissions?.[PermissionNames.permittedChains] || {};
```

This was marked resolved but I still think we should do it

NidhiKJha replied:
Updated here https://github.com/MetaMask/metamask-extension/pull/26635/commits/3868d34057e993b1cbcdf570bd5765b62e4bf5e7

NidhiKJha commented:
> if (!flattenedPermittedChains.includes(network.chainId)) {
>               dispatch(showPermittedNetworkToast());

Updated Disconnect all accounts to Disconnect and waiting for @mmragkandala to confirm what should be copy for connect
![Screenshot 2024-09-18 at 8 47 24 PM](https://github.com/user-attachments/assets/9b03d97e-3572-4c85-81bb-edad6452c1ca)

NidhiKJha committed:
nit fix

jiexi commented:
> A design/copy question: <img alt="Screenshot 2024-09-18 at 2 04 09 PM" width="360" src="https://private-user-images.githubusercontent.com/34557516/368704018-33151582-1724-465c-bfd2-0f765a725e11.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY2ODkxNDMsIm5iZiI6MTcyNjY4ODg0MywicGF0aCI6Ii8zNDU1NzUxNi8zNjg3MDQwMTgtMzMxNTE1ODItMTcyNC00NjVjLWJmZDItMGY3NjVhNzI1ZTExLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTE4VDE5NDcyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVhMDVhMjdkODU1ZDU1ZGU5ZjU0ZWQ0NzE4ZWQ1YzBmNWI2YTJhN2Q4YmQ5NmIwYmZhZWJjYzE5NWMzZWU3ZTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.KQ3jGyMynPZLty7Z8mKbPyuOU1iIbqm4Ky76P1wEFls">
> 
> Should the disconnect button now read "Disconnect all accounts and networks"?
> 
> Same question with the manual "connect accounts" button: <img alt="Screenshot 2024-09-18 at 2 05 22 PM" width="358" src="https://private-user-images.githubusercontent.com/34557516/368704257-62e8981b-98c0-4e33-b9e0-a0a3fc58a2c5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY2ODkxNDMsIm5iZiI6MTcyNjY4ODg0MywicGF0aCI6Ii8zNDU1NzUxNi8zNjg3MDQyNTctNjJlODk4MWItOThjMC00ZTMzLWI5ZTAtYTBhM2ZjNThhMmM1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTE4VDE5NDcyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNhZWI2OTMxMjBiOTE5MWNjZWRmZjRmYjEwMzQ3YjI4YTdhY2RkNjBiMGIwNDEzY2QzZGRmM2NjNmJjODRhMGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.J5satzH5I59NxjdTWrgQzWNuvxbtv5DKVGUw4wiQ-DA">
> 
> Should it be "connect accounts and networks"?

OR maybe just connect and disconnect? not sure about the exact copy, but I agree that it may make sense to update it

NidhiKJha commented:
> 
> OR maybe just connect and disconnect? not sure about the exact copy, but I agree that it may make sense to update it

@jiexi updated for Disconnect and waiting for Rag's confirmation for connect here https://github.com/MetaMask/metamask-extension/pull/26635#issuecomment-2359279049

jiexi commented on app/scripts/metamask-controller.js:
```diff
@@ -2787,6 +2788,40 @@ export default class MetamaskController extends EventEmitter {
+            this.networkController.getNetworkConfigurationByNetworkClientId(
+              currentNetworkClientIdForOrigin,
+            );
+          // if(chains.length === 0) {
+          // TODO: This particular case should also occur at the same time
+          // that eth_accounts is revoked. When eth_accounts is revoked,
+          // the networkClientId for that origin should be reset to track
+          // the globally selected network.
+          // }
```
`chains.length === 0` would occur in the case that the permmitedChains permission was removed

adonesky1 replied:
ah right ok. So why can we safely ignore it for now?

darkwing replied:
This should be addressed before merge

jiexi replied:
Ahh sorry, i explained this to Alex in slack. Here is a copy and paste of my response:

> it's not currently possible via UI to remove all permitted chains for a domain 
> unselecting all permitted chains in the edit network modal causes the update button to be replaced with a disconnect button which removes the permissions instead

darkwing replied:
Nit: Again, should this comment be removed?

NidhiKJha committed:
updated for test nets

jiexi committed:
rename getChangedOrigins to diffMap

jiexi committed:
Merge remote-tracking branch 'origin/editing-flow' into editing-flow

adonesky1 reviewed (dismissed):
LGTM!

darkwing commented on app/scripts/lib/rpc-method-middleware/handlers/ethereum-chain-utils.js:
```diff
@@ -207,7 +209,11 @@ export async function switchChain(
+        if (isAddFlow) {
+          await grantPermittedChainsPermissionIncremental([chainId]);
+        } else {
+          await requestPermittedChainsPermission([chainId]);
```
Would merging this change have an effect on prod MetaMask (not feature flagged)?

jiexi replied:
It shouldn't. That whole block is guarded by a check for the feature flag 

https://github.com/MetaMask/metamask-extension/blob/144dea496a48a00715de5c4cf16d13df00e43f3f/app/scripts/lib/rpc-method-middleware/handlers/ethereum-chain-utils.js#L201

darkwing commented on app/scripts/metamask-controller.js:
```diff
@@ -2787,6 +2788,40 @@ export default class MetamaskController extends EventEmitter {
       getPermittedAccountsByOrigin,
     );
 
+    this.controllerMessenger.subscribe(
```
Would this big change be problematic before release as well?

jiexi replied:
It shouldn't as there wouldn't be any way to add or modify the permittedChains permission, so this handler should effectively be one big noop.

darkwing reviewed (commented):
Overall looks pretty good.  I have a few questions before giving the go-ahead to merge.  There are also some comments from myself and Jony that went unanswered.

Lastly, I'm wondering if some of the data management that's appearing inside components should either be moved to a helper function or a selector.

NidhiKJha commented on ui/store/actions.ts:
```diff
@@ -1821,6 +1826,7 @@ export function removePermittedAccount(
   origin: string,
   address: string,
 ): ThunkAction<void, MetaMaskReduxState, unknown, AnyAction> {
+  console.log('removed is done');
```
Removed

NidhiKJha commented on ui/store/actions.ts:
```diff
@@ -3184,13 +3220,29 @@ export function toggleNetworkMenu() {
 }
 
-export function setAccountDetailsAddress(address: string) {
+export function setAccountDetailsAddress(address: string[]) {
```
I think we need to set an address with this action

NidhiKJha commented on ui/store/actions.ts:
```diff
@@ -3184,13 +3220,29 @@ export function toggleNetworkMenu() {
 
+export function setSelectedAccountsForDappConnection(
+  addresses: string[],
+): void {
```
Since this doesn't need a background state update, I think we return addresses here

NidhiKJha commented on yarn.lock:
```diff
@@ -14403,9 +14403,9 @@ __metadata:
   linkType: hard
 
 "caniuse-lite@npm:^1.0.30001109, caniuse-lite@npm:^1.0.30001587, caniuse-lite@npm:^1.0.30001599":
```
Removed

NidhiKJha commented on ui/store/actions.ts:
```diff
@@ -3184,13 +3220,29 @@ export function toggleNetworkMenu() {
+  };
+}
+
+export function setSelectedNetworksForDappConnection(addresses: []) {
```
Same as this one https://github.com/MetaMask/metamask-extension/pull/26635#discussion_r1765774178

jiexi committed:
permission-page-container named vars

jiexi committed:
lint

jiexi committed:
reduce type casts

jiexi committed:
Merge branch 'develop' into editing-flow

NidhiKJha commented:
> Overall looks pretty good. I have a few questions before giving the go-ahead to merge. There are also some comments from myself and Jony that went unanswered.

I have addressed all your comments but my bad forgot to reply. Had addressed Jony's comment in this message https://github.com/MetaMask/metamask-extension/pull/26635#issuecomment-2339006359 but addressed them separately as well.

> Lastly, I'm wondering if some of the data management that's appearing inside components should either be moved to a helper function or a selector.

I have tried to create selectors for most of the functions that are used in multiple components. But can take a look and move the data management to selectors

jiexi committed:
cleanup checked isIndeterminate assignment

NidhiKJha committed:
updated permissions page

jiexi committed:
swap !checked condition order

jiexi committed:
Merge branch 'develop' into editing-flow

jiexi committed:
Merge remote-tracking branch 'origin/editing-flow' into editing-flow

NidhiKJha committed:
lint fix

jiexi committed:
Fix approvedAccounts selection in connection flow

jiexi committed:
Merge remote-tracking branch 'origin/editing-flow' into editing-flow

jiexi committed:
Merge branch 'develop' into editing-flow

jiexi committed:
add SiteCellConnectionListItem

NidhiKJha committed:
Merge branch 'develop' into editing-flow

jiexi committed:
Add PermissionsHeader

jiexi committed:
Merge remote-tracking branch 'origin/editing-flow' into editing-flow

jiexi committed:
Fix missingi mports

NidhiKJha committed:
updated connect page

jiexi committed:
Merge branch 'develop' into editing-flow

# Conflicts:
#	app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js
#	app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.test.js
#	ui/components/multichain/edit-networks-modal/edit-networks-modal.js
#	ui/components/multichain/network-list-menu/network-list-menu.js

jiexi committed:
Jl/editing flow refactor (#27309)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/27309?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask
Extension Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

NidhiKJha committed:
story fix site cell

NidhiKJha committed:
fixed story for accounts

NidhiKJha committed:
fixed test for edit networks

NidhiKJha committed:
updated lint fix

NidhiKJha committed:
Merge branch 'develop' into editing-flow

NidhiKJha committed:
updated copy for disconnect modal

NidhiKJha committed:
lint fix

jiexi committed:
Lint

jiexi committed:
Merge branch 'develop' into editing-flow

jiexi committed:
Shallow review-permissions-page connect-page snapshot test

jiexi committed:
Rename Request type. lint

jiexi committed:
Merge branch 'develop' into editing-flow

jiexi committed:
Shallow EditNetworksModal test

jiexi committed:
add actions specs

jiexi committed:
lint

jiexi committed:
Switch actions.test.js back to sinon

jiexi committed:
Merge branch 'develop' into editing-flow

jiexi committed:
add background-api spec

jiexi committed:
update background-api test requestAccountsAndChainPermissionsWithId

jiexi commented on ui/pages/routes/routes.component.js:
```diff
@@ -759,6 +771,31 @@ export default class Routes extends Component {
+              <AvatarNetwork
+                size={AvatarAccountSize.Md}
+                borderColor={BorderColor.transparent}
+                src={currentNetwork?.rpcPrefs.imageUrl || ''}
```
this may be incorrect after the network upgrade merge

jiexi replied:
this is fine because of massaging in the `getCurrentNetwork` selector

jiexi committed:
Add edit modal scenarios

NidhiKJha committed:
Merge branch 'develop' into editing-flow

NidhiKJha committed:
lint fix

NidhiKJha committed:
lint fix

NidhiKJha committed:
updated test for site cell

NidhiKJha committed:
updated test

NidhiKJha committed:
updated test and fixed prep dep

NidhiKJha committed:
lintfix

jiexi committed:
Fix restricted import in actions test

jiexi committed:
add EndowmentTypes to shared constants

jiexi committed:
Fix actions

jiexi committed:
Merge branch 'develop' into editing-flow

jiexi committed:
Remove infura key

jiexi committed:
Merge branch 'develop' into editing-flow

jiexi committed:
Fix wrong return type in actions

jiexi committed:
rename addMorePermittedChains

jiexi committed:
Fix review permissions permitted chains

jiexi committed:
Fix background-api test

jiexi committed:
Fix action return types

darkwing commented on ui/components/multichain/pages/connections/connections.tsx:
```diff
@@ -206,50 +192,10 @@ export const Connections = () => {
-      </Header>
+      <PermissionsHeader
+        securedOrigin={securedOrigin}
+        connectedSubjectsMetadata={connectedSubjectsMetadata}
```
Is this a related change?

NidhiKJha replied:
Nice catch, no! this file shouldn't be changed. Updated https://github.com/MetaMask/metamask-extension/pull/26635/commits/76808f645e610095ca5c83e83c9d76445db545db

adonesky1 commented:
Seeing an issue where the network avatar/icon isn't showing up correctly for the adding a single chain permission on `wallet_switchEthereumChain` calls:

https://github.com/user-attachments/assets/8486494c-dfdc-4023-bf2e-c74d353b65f3

darkwing commented on app/scripts/controllers/permissions/selectors.test.js:
```diff
@@ -113,4 +117,89 @@ describe('PermissionController selectors', () => {
   });
+
+  describe('getPermittedChainsByOrigin', () => {
+    it('memoizes and gets permitted chains by origin', () => {
```
Nice work adding tests here!

adonesky1 commented:
Seeing an issue when switching to a network not currently permissioned for a dapp:
1) I don't see the toast indicating the permission is granted unless on the home scree, so I don't see it on the permssions view
2) the permission is no longer actually being granted:

https://github.com/user-attachments/assets/8959b481-2357-4d0f-89ab-27706cef0e03

darkwing commented on ui/pages/permissions-connect/connect-page/connect-page.tsx:
```diff
@@ -0,0 +1,149 @@
+  id: string;
+  origin: string;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  [key: string]: any; // Adjust this if you have a more specific shape for the request
```
Shouldn't we know what type this is?

darkwing commented on ui/components/multichain/global-menu/global-menu.js:
```diff
@@ -253,7 +253,6 @@ export const GlobalMenu = ({ closeMenu, anchorElement, isOpen }) => {
       >
         {t('allPermissions')}
       </MenuItem>
-
```
Not needed, can remove to make PR smaller.

NidhiKJha replied:
Added back https://github.com/MetaMask/metamask-extension/pull/26635/commits/9b8a4843e14e13f22607be4ca864747d524a1948

darkwing commented on ui/components/multichain/pages/review-permissions-page/site-cell/site-cell.stories.tsx:
```diff
@@ -0,0 +1,95 @@
+
+export const DefaultStory = (args) => <SiteCell {...args} />;
+
+DefaultStory.storyName = 'Default';
```
Needs new line

darkwing commented:
Left a few comments; in the future, we really need to isolate our feature PRs to the issue they represent, otherwise we risk large, hard to review PRs that take ages to get merged.

NidhiKJha committed:
added logo to review page

NidhiKJha commented:
> Seeing an issue where the network avatar/icon isn't showing up correctly for the adding a single chain permission on `wallet_switchEthereumChain` calls:
> 
Nice catch! Updated here  https://github.com/MetaMask/metamask-extension/pull/26635/commits/4bdf4bcddbff9e854547fcb713137680feae2bf3
![Screenshot 2024-09-25 at 1 49 25 PM](https://github.com/user-attachments/assets/8d3a80e2-9502-47ee-b428-98f20de84897)

NidhiKJha committed:
removed unnecessary type

NidhiKJha committed:
nit fix

NidhiKJha committed:
lint fix

NidhiKJha committed:
reverted change for connections

NidhiKJha committed:
updated site cell

NidhiKJha commented:
> The hover state over the whole row seems to imply you can click but you can't anywhere accept the three dot menu button. Is this intended?

No, only button should be clickable and avatar group should on tooltip. Updated here https://github.com/MetaMask/metamask-extension/pull/26635/commits/0342f157d1e3996aff15d5609ceda39e4337e4cc

NidhiKJha commented:
> Seeing an issue where the network avatar/icon isn't showing up correctly for the adding a single chain permission on `wallet_switchEthereumChain` calls:
Switch Modal should show up only on home page not on Permissions Page. 

Looking into why the networks are not updated 👀

jiexi committed:
fix typo in actions

jiexi committed:
Merge remote-tracking branch 'origin/editing-flow' into editing-flow

jiexi committed:
Merge branch 'develop' into editing-flow

jiexi commented:
> Seeing an issue when switching to a network not currently permissioned for a dapp:

Great catch. Should be fixed here https://github.com/MetaMask/metamask-extension/pull/26635/commits/1bc9e0f0cb62f8a2432ac19a0f2bbccd6c634436

https://github.com/user-attachments/assets/3a73e5b6-fe48-45bd-979b-7404de323d99

NidhiKJha committed:
updated test

NidhiKJha committed:
updated test

jiexi committed:
Fix action grantPermittedChain specs

jiexi committed:
Merge branch 'develop' into editing-flow

jiexi committed:
Merge remote-tracking branch 'origin/editing-flow' into editing-flow

jiexi committed:
do not grant permitted changes on network switch if origin does not have permitted accounts

jiexi committed:
fix permitted chain toast being shown when network is switched to an unpermitted chain on the review permission component on permitted origin and then the origin is disconnected

jiexi committed:
Use showPermittedNetworkToastOpen in ReviewPermissions page to trigger setShowNetworkToast

NidhiKJha committed:
Merge branch 'develop' into editing-flow

adonesky1 reviewed (approved):
Awesome! So excited for this to go live!

Mrtenz commented on ui/store/actions.ts:
```diff
@@ -1811,6 +1815,67 @@ export function removePermittedAccount(
+    });
+    await forceUpdateMetamaskState(dispatch);
+  };
+}
```
Nit:
```suggestion
}

```

Mrtenz reviewed (approved):
Snaps-related changes look good. I haven't reviewed everything else in-depth.

darkwing commented on ui/pages/permissions-connect/connect-page/connect-page.tsx:
```diff
@@ -0,0 +1,147 @@
+        [[] as NetworkConfiguration[], [] as NetworkConfiguration[]],
+      ),
+    [networkConfigurations],
+  );
```
Nit: Please move this to maybe a selector or somewhere outside of this UI component if possible.

darkwing reviewed (approved):
Nit:  The Sepolia network circle doesn't have the same background color as the networks menu:

<img width="355" alt="SCR-20240926-jnre" src="https://github.com/user-attachments/assets/be3e807e-9873-49a0-b42b-a5102977d744">
